proxy/src/main/java/com/velocitypowered/proxy/provider/ClickCallbackProviderImpl.java

```diff
@@ -0,0 +1,139 @@
+      final @NotNull ClickCallback<Audience> callback,
+      final ClickCallback.@NotNull Options options) {
+    return ClickEvent.runCommand(
+        "/velocity callback " + CALLBACK_MANAGER.addCallback(callback, options));
+  }
+
+  /** The callback manager. */
+  public static final class CallbackManager {
+
+    private final Map<UUID, StoredCallback> callbacks = new HashMap<>();
+    private final Queue<StoredCallback> queue = new ConcurrentLinkedQueue<>();
+
+    private CallbackManager() {}
+
+    /**
+     * Add a callback to the queue.
+     *
+     * @param callback the callback to add to the queue
+     * @param options callback options
+     * @return the {@link UUID} of the callback
+     */
+    public UUID addCallback(
+        final @NotNull ClickCallback<Audience> callback,
+        final ClickCallback.@NotNull Options options) {
+      final UUID id = UUID.randomUUID();
+      this.queue.add(new StoredCallback(callback, options, id));
+      return id;
+    }
+
+    /**
+     * Handle queue.
+     */
+    public void handleQueue() {
+      // Evict expired entries
+      this.callbacks.values().removeIf(callback -> !callback.valid());
+
+      // Add entries from queue
+      StoredCallback callback;
+      while ((callback = this.queue.poll()) != null) {
+        this.callbacks.put(callback.id(), callback);
+      }
+    }
+
+    /**
+     * Run a callback.
+     *
+     * @param audience the audience
+     * @param id the callbacks id
+     */
+    public void runCallback(final @NotNull Audience audience, final UUID id) {
+      final StoredCallback callback = this.callbacks.get(id);
+      if (callback != null && callback.valid()) { // TODO Message if expired/invalid?
+        callback.takeUse();
+        callback.callback.accept(audience);
+      }
+    }
+  }
+
+  /** A stored callback. */
+  private static final class StoredCallback {
+    private final long startedAt = System.nanoTime();
+    private final ClickCallback<Audience> callback;
+    private final long lifetime;
+    private final UUID id;
+    private int remainingUses;
+
+    private StoredCallback(
+        final @NotNull ClickCallback<Audience> callback,
+        final ClickCallback.@NotNull Options options,
+        final UUID id) {
+      this.callback = callback;
+      this.lifetime = options.lifetime().toNanos();
+      this.remainingUses = options.uses();
+      this.id = id;
+    }
+
+    public void takeUse() {
+      if (this.remainingUses != ClickCallback.UNLIMITED_USES) {
+        this.remainingUses--;
+      }
+    }
+
+    public boolean hasRemainingUses() {
+      return this.remainingUses == ClickCallback.UNLIMITED_USES || this.remainingUses > 0;
+    }
+
+    public boolean expired() {
+      return System.nanoTime() - this.startedAt >= this.lifetime;
+    }
+
+    public boolean valid() {
+      return hasRemainingUses() && !expired();
+    }
+
+    public UUID id() {
+      return this.id;
+    }
+  }
```

Comment:
I'm really not a big fan of this approach.

* We already have an expiring cache, Caffeine. Why have an eviction task running every 5 milliseconds?
* The `HashMap` used here is not thread safe. Velocity is a multi-threaded environment. This kind of implementation would make sense with Paper, with a single server thread, but it won't fly here.

For this reason I am not willing to sign off on this implementation unless these issues were resolved.